### PR TITLE
Swap C# completion and parameter hinting engine to roslyn.

### PR DESCRIPTION
**Pull request by mkrueger**

### CODE REVIEW

**Comment by mkrueger** [comment]
It's a work in progress - not done yet.

File: main/src/addins/CSharpBinding/MonoDevelop.CSharp.Completion/CSharpCompletionTextEditorExtension.cs [hunk 1]
```diff
@@ -417,68 +410,78 @@ internal void AddImportCompletionData (CompletionResult completionResult, CSharp
...
-					list.DefaultCompletionString = completionResult.DefaultCompletionString;
-					// list.CloseOnSquareBrackets = completionResult.CloseOnSquareBrackets;
-					list.AutoCompleteUniqueMatch = triggerInfo.CompletionTriggerReason == CompletionTriggerReason.CompletionCommand;
-				} catch (OperationCanceledException) {
-					return null;
-				} catch (AggregateException e) {
-					e.Flatten ().Handle (f => f is OperationCanceledException);
-				} catch (Exception e) {
-					LoggingService.LogError ("Error while getting C# recommendations", e); 
+			var cs = DocumentContext.RoslynWorkspace.Services.GetLanguageServices (LanguageNames.CSharp).GetService<CompletionService> ();
```
Comment (Therzok) [inline main/src/addins/CSharpBinding/MonoDevelop.CSharp.Completion/CSharpCompletionTextEditorExtension.cs]: `Microsoft.CodeAnalysis.Shared.Extensions` has an extension method for `GetLanguageService (this Document)`

File: main/src/addins/CSharpBinding/MonoDevelop.CSharp.Completion/RoslynParameterHintingEngine.cs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+using System.Collections.Concurrent;
+using MonoDevelop.Core;
+
+namespace MonoDevelop.CSharp.Completion
+{
+	class RoslynParameterHintingEngine
+	{
+		public async Task<ParameterHintingResult> GetParameterDataProviderAsync (Document document, int position, CancellationToken token = default (CancellationToken))
+		{
+			var providers = CSharpCompletionTextEditorExtension.signatureProviders.Value.ToList ();
```
Comment (Therzok) [inline main/src/addins/CSharpBinding/MonoDevelop.CSharp.Completion/RoslynParameterHintingEngine.cs]: The list of providers will never change, MEF composition is done at startup?

File: main/src/addins/CSharpBinding/MonoDevelop.CSharp.Completion/RoslynParameterHintingEngine.cs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+			var triggerInfo = new SignatureHelpTriggerInfo (SignatureHelpTriggerReason.TypeCharCommand, (await document.GetTextAsync (token)).ToString ()[position - 1]);
```
Comment (Therzok) [inline main/src/addins/CSharpBinding/MonoDevelop.CSharp.Completion/RoslynParameterHintingEngine.cs]: Can we _not_ tostring the whole document text? We can pass in a span here.

File: main/src/addins/CSharpBinding/MonoDevelop.CSharp.Completion/RoslynParameterHintingEngine.cs [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+			return await GetParameterDataProviderAsync (providers,  document, position, triggerInfo, token);
+		}
+
+		public async Task<ParameterHintingResult> GetParameterDataProviderAsync (List<ISignatureHelpProvider> providers, Document document, int position, SignatureHelpTriggerInfo triggerInfo, CancellationToken token = default (CancellationToken))
+		{
+			var hintingData = new ConcurrentBag<ParameterHintingData> ();
```
Comment (Therzok) [inline main/src/addins/CSharpBinding/MonoDevelop.CSharp.Completion/RoslynParameterHintingEngine.cs]: ConcurrentBag? The whole code below is serial.

File: main/src/addins/CSharpBinding/MonoDevelop.CSharp.Completion/SignatureHelpParameterHintingData.cs [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+		public override bool Equals (object obj)
+		{
+			var other = obj as SignatureHelpParameterHintingData;
+			if (other == null)
+				return false;
+			return Item.ToString () == other.Item.ToString ();
+		}
+
+		public override Task<TooltipInformation> CreateTooltipInformation (TextEditor editor, DocumentContext ctx, int currentParameter, bool smartWrap, CancellationToken cancelToken)
```
Comment (Therzok) [inline main/src/addins/CSharpBinding/MonoDevelop.CSharp.Completion/SignatureHelpParameterHintingData.cs]: :heart: I'm going to use some bits of this.

File: main/src/addins/CSharpBinding/MonoDevelop.CSharp.Refactoring/CSharpSnippetInfoService.cs [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+namespace MonoDevelop.CSharp.Refactoring
+{
+	[ExportLanguageService (typeof (ISnippetInfoService), LanguageNames.CSharp), Shared]
+	class CSharpSnippetInfoService : ISnippetInfoService
+	{
+		IEnumerable<SnippetInfo> ISnippetInfoService.GetSnippetsIfAvailable ()
+		{
+			foreach (var template in CodeTemplateService.GetCodeTemplates (CSharp.Formatting.CSharpFormatter.MimeType)) {
+				yield return new SnippetInfo (template.Shortcut, template.Shortcut, template.Description, "Path!");
```
Comment (Therzok) [inline main/src/addins/CSharpBinding/MonoDevelop.CSharp.Refactoring/CSharpSnippetInfoService.cs]: What is `"Path!"`?

File: main/src/addins/CSharpBinding/MonoDevelop.CSharp/SignatureMarkupCreator.cs [hunk 7]
```diff
@@ -148,17 +148,35 @@ public string GetTypeReferenceString (ITypeSymbol type, bool highlight = true)
...
 
-			if (ICSharpCode.NRefactory6.CSharp.Completion.ObjectCreationContextHandler.primitiveTypesKeywords.Contains (displayString))
+			if (primitiveTypesKeywords.Contains (displayString))
 				return Highlight (displayString, GetThemeColor (keywordType));
 			var text = MonoDevelop.Ide.TypeSystem.Ambience.EscapeText (displayString);
 			return highlight ? HighlightSemantically (text, GetThemeColor (userTypes)) : text;
 		}
 
+		static string [] primitiveTypesKeywords = new string [] {
+			"void",
```
Comment (Therzok) [inline main/src/addins/CSharpBinding/MonoDevelop.CSharp/SignatureMarkupCreator.cs]: Please use http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp/Syntax/SyntaxKindFacts.cs,1201 on the syntax kinds.

You can get the c# language service for Syntax Facts

File: main/src/addins/CSharpBinding/MonoDevelop.CSharp/SignatureMarkupCreator.cs [hunk 8]
```diff
@@ -148,17 +148,35 @@ public string GetTypeReferenceString (ITypeSymbol type, bool highlight = true)
...
 				//but in case of Sketches where user usually is at end of document when typing text this can throw exception
 				//because offset can be >= Length
-				displayString = model != null ? RoslynCompletionData.SafeMinimalDisplayString (type, model, Math.Min (model.SyntaxTree.Length - 1, offset), MonoDevelop.Ide.TypeSystem.Ambience.LabelFormat) : type.Name;
+				displayString = model != null ? CSharpAmbience.SafeMinimalDisplayString (type, model, Math.Min (model.SyntaxTree.Length - 1, offset), MonoDevelop.Ide.TypeSystem.Ambience.LabelFormat) : type.Name;
 			} else {
 				displayString = type.ToDisplayString (MonoDevelop.Ide.TypeSystem.Ambience.LabelFormat);
 			}
 
-			if (ICSharpCode.NRefactory6.CSharp.Completion.ObjectCreationContextHandler.primitiveTypesKeywords.Contains (displayString))
+			if (primitiveTypesKeywords.Contains (displayString))
```
Comment (Therzok) [inline main/src/addins/CSharpBinding/MonoDevelop.CSharp/SignatureMarkupCreator.cs]: Please use http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp/Syntax/SyntaxKindFacts.cs,1201 on the syntax kinds.

You can get the c# language service for Syntax Facts

**Comment by Therzok** [comment]
Done with my initial review pass.

File: main/src/addins/CSharpBinding/CSharpBinding.csproj [hunk 9]
```diff
@@ -450,5 +419,6 @@
   <ItemGroup>
     <Folder Include="icons\" />
     <Folder Include="MonoDevelop.CSharp.NRefactoryWrapper\" />
+    <Folder Include="MonoDevelop.CSharp.Completion\CompletionProvider\" />
```
Comment (KirillOsenkov) [inline main/src/addins/CSharpBinding/CSharpBinding.csproj]: ><Folder Include="MonoDevelop.CSharp.Completion\CompletionProvider\" /> [](start = 4, length = 70)

No need to include an empty folder?

File: main/src/addins/CSharpBinding/MonoDevelop.CSharp.Completion/CSharpCompletionTextEditorExtension.cs [hunk 10]
```diff
@@ -317,15 +310,15 @@ interface IListData
 		}
 
 
-		internal void AddImportCompletionData (CompletionResult completionResult, CSharpCompletionDataList result, RoslynCodeCompletionFactory factory, SemanticModel semanticModel, int position, CancellationToken cancellationToken = default(CancellationToken))
+		internal void AddImportCompletionData (CSharpSyntaxContext ctx, CompletionDataList result, SemanticModel semanticModel, int position, CancellationToken cancellationToken = default(CancellationToken))
```
Comment (KirillOsenkov) [inline main/src/addins/CSharpBinding/MonoDevelop.CSharp.Completion/CSharpCompletionTextEditorExtension.cs]: >ctx [](start = 61, length = 3)

Avd abbrvtns

File: main/src/addins/CSharpBinding/MonoDevelop.CSharp.Completion/RoslynCompletionData.cs [hunk 11]
```diff
@@ -23,96 +23,242 @@
...
+					result = DisplayFlags.MarkedBold;
+				}
+
+				return result;
+			}
+		}
+
+		public override async Task<TooltipInformation> CreateTooltipInformation (bool smartWrap, CancellationToken cancelToken)
+		{
+			var tt = new TooltipInformation ();
```
Comment (KirillOsenkov) [inline main/src/addins/CSharpBinding/MonoDevelop.CSharp.Completion/RoslynCompletionData.cs]: >tt [](start = 7, length = 2)

tooltipInformation

File: main/src/addins/CSharpBinding/MonoDevelop.CSharp.Completion/RoslynCompletionData.cs [hunk 12]
```diff
@@ -23,96 +23,242 @@
...
+			var description = await completionService.GetDescriptionAsync (doc, CompletionItem);
+			var markup = new StringBuilder ();
+			var theme = DefaultSourceEditorOptions.Instance.GetEditorTheme ();
+			var tp = description.TaggedParts;
```
Comment (KirillOsenkov) [inline main/src/addins/CSharpBinding/MonoDevelop.CSharp.Completion/RoslynCompletionData.cs]: >tp [](start = 7, length = 2)

taggedParts

File: main/src/addins/CSharpBinding/MonoDevelop.CSharp.Completion/SignatureHelpParameterHintingData.cs [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+using System.Linq;
+using MonoDevelop.Ide.Fonts;
+
+namespace MonoDevelop.CSharp.Completion
+{
+	class SignatureHelpParameterHintingData : Ide.CodeCompletion.ParameterHintingData
+	{
+		public SignatureHelpParameterHintingData (SignatureHelpItem item)
+		{
+			Item = item;
```
Comment (KirillOsenkov) [inline main/src/addins/CSharpBinding/MonoDevelop.CSharp.Completion/SignatureHelpParameterHintingData.cs]: >item [](start = 10, length = 4)

Is item allowed to be null?

File: main/src/addins/CSharpBinding/MonoDevelop.CSharp.Highlighting/HighlightUsagesExtension.cs [hunk 14]
```diff
@@ -78,7 +78,7 @@ static HighlightUsagesExtension ()
 					.ToArray ();
 			} catch (Exception e) {
 				LoggingService.LogError ("Error while loading highlighters.", e);
-
+				highlighters = new IHighlighter [0];
```
Comment (KirillOsenkov) [inline main/src/addins/CSharpBinding/MonoDevelop.CSharp.Highlighting/HighlightUsagesExtension.cs]: >new IHighlighter [0] [](start = 19, length = 20)

Array.Empty<IHighlighter>()

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MetadataReferenceCache.cs [hunk 15]
```diff
@@ -184,6 +187,31 @@ void CreateNewReference ()
 					}
 				}
 			}
+
+
+			class RoslynDocumenentationProvider : DocumentationProvider
+			{
+				internal static readonly DocumentationProvider Instance = new RoslynDocumenentationProvider ();
+
+				RoslynDocumenentationProvider ()
```
Comment (KirillOsenkov) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MetadataReferenceCache.cs]: >e [](start = 17, length = 1)

Typo: Documenentation -> Documentation

File: main/src/addins/CSharpBinding/MonoDevelop.CSharp.Completion/RoslynParameterHintingEngine.cs (hunk 2 above)
Comment (mkrueger) [inline main/src/addins/CSharpBinding/MonoDevelop.CSharp.Completion/RoslynParameterHintingEngine.cs]: You implemented that :)

File: main/src/addins/CSharpBinding/MonoDevelop.CSharp.Refactoring/CSharpSnippetInfoService.cs (hunk 6 above)
Comment (mkrueger) [inline main/src/addins/CSharpBinding/MonoDevelop.CSharp.Refactoring/CSharpSnippetInfoService.cs]: I've no idea - still need to figure that out.

File: main/src/addins/CSharpBinding/CSharpBinding.csproj (hunk 9 above)
Comment (mkrueger) [inline main/src/addins/CSharpBinding/CSharpBinding.csproj]: The IDE does that

File: main/src/addins/CSharpBinding/MonoDevelop.CSharp.Completion/CSharpCompletionTextEditorExtension.cs (hunk 10 above)
Comment (mkrueger) [inline main/src/addins/CSharpBinding/MonoDevelop.CSharp.Completion/CSharpCompletionTextEditorExtension.cs]: ntrnl

**Comment by KirillOsenkov** [comment]
@mkrueger Mike, could you rebase and also add a build lane so we can see if it builds and tests pass on CI?

**Pull request merged**